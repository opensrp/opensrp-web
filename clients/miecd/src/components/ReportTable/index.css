.patient-details tbody tr {
  display: table-row;
}

.patient-details #tableRow {
  margin: 10px 0px 50px 0;
  background-color: #ffffff;
  border: 1px solid #dadce0;
  border-radius: 5px;
  box-shadow: 0 1px 3px 0 #d7d7e0;
}

.patient-details #dropdownRow {
  width: 100%;
}

.patient-details .dropdown button span {
  color: #676767;
  font-size: 13px;
  font-weight: 500;
  line-height: 16px;
}

.patient-details #dropdown {
  background-color: #ffffff;
  color: #666666;
  border: 0.5px solid #bababa;
  border-radius: 2px;
  box-shadow: 0 1px 3px 1px rgba(127, 131, 155, 0.15), 0 1px 2px 0 rgba(127, 131, 155, 0.3);
}

.patient-details .table-container thead tr th:not(:first-child) {
  border-left: 1px solid #e8e8e8;
}

.patient-details .filter-panel {
  display: flex;
  flex-direction: row;
  flex-wrap: nowrap;
  justify-content: flex-start;
}

.patient-details .filter-panel .filters {
  display: flex;
  justify-content: flex-start;
  margin-left: 3%;
  flex-grow: 2;
  padding: 0 0 20px;
}

.patient-details thead {
  position: relative;
  top: 0;
  height: 70px;
}

.patient-details td {
  width: 25%;
  border: 1px solid #e8e8e8;
}

.patient-details .table-container {
  width: 100%;
  box-shadow: initial;
  height: auto;
}

.patient-details td:nth-child(2n) {
  background-color: #fcfcfc;
}

.patient-details td:nth-child(2n + 1) {
  background-color: #ffffff;
}

.patient-details th:nth-child(2n) {
  background-color: #f5f5f5;
}

.patient-details th:nth-child(2n + 1) {
  background-color: #fafafa;
}

.patient-details #dropdownRow > p {
  color: #001429;
}

.patient-details #dropdown-basic {
  padding-top: 0;
  margin-top: -0.2rem;
  height: 40px;
}

.patient-details #dropdown-basic.dropdown.dropdown-toggle,
.patient-details #dropdown-basic:hover {
  color: #666666;
  box-shadow: none;
  color: black;
  background-color: #ffffff;
  border: 0.5px solid #bababa;
  border-radius: 2px;
  box-shadow: 0 1px 3px 1px rgba(127, 131, 155, 0.15), 0 1px 2px 0 rgba(127, 131, 155, 0.3);
}

.patient-details th {
  color: #001429;
  font-size: 16px;
  font-weight: 400;
  letter-spacing: -0.18px;
  line-height: 19px;
  text-align: left;
  padding-left: 20px;
}

.patient-details td {
  color: #334354;
  font-size: 15px;
  font-weight: 400;
  line-height: 18px;
  text-align: left;
  padding: 20px;
}
