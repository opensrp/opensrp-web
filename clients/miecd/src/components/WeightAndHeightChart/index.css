div#chart-wrapper {
  margin: auto;
}

.highcharts-point {
  fill: white;
  stroke: #0093db;
  stroke-width: 2px;
}

#chart > .card {
  margin-bottom: 30px;
  width: 100%;
}
