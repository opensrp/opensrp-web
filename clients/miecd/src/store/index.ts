import { connectReducer, getConnectedStore } from '@onaio/connected-reducer-registry';
import { gateKeeperReducer, gateKeeperReducerName } from '@onaio/gatekeeper';
import reducerRegistry, { combine, Registry } from '@onaio/redux-reducer-registry';
import session, { reducerName as sessionReducer } from '@onaio/session-reducer';

/** Initial reducers in the reducer registry */
const defaultReducers: Registry = {
    router: connectReducer as any /** Dirty hack because connectRouter LocationChangeAction does not extend Redux.AnyAction */,
};

/** Add users reducer to registry */
defaultReducers[sessionReducer] = session;
defaultReducers[gateKeeperReducerName] = gateKeeperReducer;

/** The initial store for the reveal web app */
const store = getConnectedStore(defaultReducers);

/** Set listener to add reducers to store when registered */
reducerRegistry.setChangeListener(reducers => {
    store.replaceReducer(combine(reducers));
});

export default store;
