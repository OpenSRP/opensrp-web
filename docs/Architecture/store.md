# overview

This documentation will take you through the process of creating and working with:

    - reducers
    - actions
    - selectors

## Where do I find the store

the store reducer modules are found within src/store. The initial directory structure looks something like this: (Note: this structure only depicts the initial directory structure and will as expected contain more directories and files as more are added with time)

```
...
├── src
│   ├── ...
│   ├── store
│   │   ├── ducks
│   │   │    ├── superset.ts
│   │   │    ├── tests/
│   │   ├── tests
│   │   ├── index.ts
│   │   └── selectors.ts
...
```

A single reducer module will contain the afore mentioned parts i.e. its respective actions, action creators if any, its selectors and the reducer function

## Actions

An action can be any object with a type property, they are the only way provisioned for changing the state in a redux store.

### Adding Actions

Its a common convention as well as recommended one to Use type constants, [read more on this here]().

steps:

- Add An interface describing the action structure: redux exposes an `AnyAction` type that can be extended to write custom action interfaces
- Create a general actiontypes to be used to type check all actions that will be passed to the reducer function as an argument. this will be in the form of [intersection type]()

## Selectors

whereas the action puts data in the store; the selectors on the other hand act as getters of the slice of data that this reducer module handles

selectors are functions that will take at least one argument, the store, Then based on any other arguments like say an id, can be used to filter down the data retrieved.

## The Reducer

The reducer function, takes 2 arguments; the state and an action.
using the type property of the action, the reducer should then know how to modify the state.
such functionality could be implemented as a form of switch strcture such as a list of if statements or the more recommended approach which is to use the actual switch statement

### Adding the Reducer

- Define an interface for the data to be stored in the store
- Define an interface for the store state.
- write the actual reducer function, while applying the required type annotations as required.
