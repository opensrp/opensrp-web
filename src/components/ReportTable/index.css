tbody tr {
  display: table-row;
}

#tableRow {
  width: 84%;
  background-color: #ffffff;
  border: 1px solid #dadce0;
  border-radius: 5px;
  box-shadow: 0 1px 3px 0 #d7d7e0;
}

#dropdownRow {
  width: 84%;
}

#dropdown {
  background-color: #ffffff;
  border: 0.5px solid #bababa;
  border-radius: 2px;
  box-shadow: 0 1px 3px 1px rgba(127, 131, 155, 0.15), 0 1px 2px 0 rgba(127, 131, 155, 0.3);
}

thead {
  position: relative;
  top: 0;
  height: 50px;
  border-bottom: 0.5px solid rgba(0, 0, 0, 0.2);
}

td {
  width: 25%;
}

.table-container {
  width: 100%;
}

path.highcharts-tracker-line {
  fill: #0093db;
}

td:nth-child(2n),
th:nth-child(2n) {
  background-color: #fcfcfc;
  border: 1px solid #e8e8e8;
}
