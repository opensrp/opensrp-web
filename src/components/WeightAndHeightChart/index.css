div#chart-wrapper {
  margin: auto;
}

.card {
  margin: auto;
}

.highcharts-point {
  fill: white;
  stroke: #0093db;
  stroke-width: 2px;
}

#chart > .card {
  margin-bottom: 30px;
}
