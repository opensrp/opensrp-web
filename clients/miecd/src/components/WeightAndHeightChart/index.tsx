/* eslint-disable @typescript-eslint/no-this-alias */
import * as Highcharts from 'highcharts';
import * as React from 'react';
import { Card, CardTitle } from 'reactstrap';
import { monthNames } from '../../constants';
import { FlexObject } from '../../helpers/utils';
import { WeightMonthYear } from '../ReportTable';
import './index.css';

interface Props {
    weights: WeightMonthYear[];
    chartWrapperId: string;
    title: string;
    units: string;
    legendString: string;
    xAxisLabel: string;
}

interface State {
    chart: any;
}

const defaultProps: Props = {
    chartWrapperId: '',
    legendString: '',
    title: '',
    units: '',
    weights: [],
    xAxisLabel: '',
};
export default class WeightAndHeightChart extends React.Component<Props, State> {
    public static defaultProps = defaultProps;
    public static legendString = defaultProps.legendString;
    public static units = defaultProps.units;
    public static xAxisLabel = defaultProps.xAxisLabel;
    public static getDerivedStateFromProps(nextProps: Props, prevState: State) {
        if (prevState.chart && prevState.chart.series) {
            return {
                chart: prevState.chart.series[0].setData(nextProps.weights),
            };
        } else {
            return {
                chart: prevState.chart,
            };
        }
    }
    public static calcChartWidth(chart: any) {
        if (chart && chart.series) {
            chart.setSize(0.8 * window.innerWidth, null);
            return chart;
        }
    }

    constructor(props: Props) {
        super(props);
        // Init state.
        this.state = { chart: null };
    }

    public componentDidMount(): void {
        let chart: any;
        const self: FlexObject = this;
        if (self.state.chart) {
            self.state.chart.destroy();
        }
        self.timeout = setTimeout(() => {
            chart = Highcharts.chart(`${this.props.chartWrapperId}`, {
                chart: {
                    type: 'line',
                    width: 0.8 * window.innerWidth,
                },
                legend: {
                    align: 'right',
                    layout: 'vertical',
                    verticalAlign: 'middle',
                },
                title: {
                    text: '',
                },
                // tslint:disable-next-line: object-literal-sort-keys
                tooltip: {
                    backgroundColor: 'white',
                    borderColor: '#DADCE0',
                    borderRadius: 10,
                    borderWidth: 1,
                    shadow: {
                        color: '#D7D7E0',
                        offsetX: 0,
                        offsetY: 2,
                        opacity: 0.2,
                        width: 8,
                    },
                    formatter() {
                        return `${self.props.legendString} - ${(this.x + '').slice(
                            0,
                            (this.x + '').lastIndexOf(' '),
                        )}<br>${self.props.xAxisLabel}  <b>${this.y}</b> ${self.props.units}`;
                    },
                },

                subtitle: {
                    text: '',
                },

                yAxis: {
                    title: {
                        text: '',
                    },
                },

                xAxis: {
                    categories: this.props.weights
                        .map((weightMonthYear: WeightMonthYear) => weightMonthYear.month)
                        .map((month: number) => monthNames[month])
                        .map((mothName: string, index: number) => `${mothName} ${this.props.weights[index].year}`),
                },

                series: [
                    {
                        data: this.props.weights.map((weightMonthYear: WeightMonthYear) => weightMonthYear.weight),
                        name: this.props.xAxisLabel,
                    },
                ] as any,

                responsive: {
                    rules: [
                        {
                            chartOptions: {
                                legend: {
                                    align: 'center',
                                    layout: 'horizontal',
                                    verticalAlign: 'bottom',
                                },
                            },
                            condition: {
                                maxWidth: 5000,
                            },
                        },
                    ],
                },
            });
        }, 300);
        window.addEventListener('resize', () => WeightAndHeightChart.calcChartWidth(chart));
        this.setState({ chart });
    }

    public componentWillUnmount(): void {
        const self: FlexObject = this;
        if (self.state.chart) {
            window.removeEventListener('resize', () => WeightAndHeightChart.calcChartWidth(this.state.chart));
            clearTimeout(self.timeout);
        }
    }

    public render() {
        return (
            <Card>
                <CardTitle>{this.props.title}</CardTitle>
                <div id={this.props.chartWrapperId} />
            </Card>
        );
    }
}
