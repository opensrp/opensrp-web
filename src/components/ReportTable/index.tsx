import ListView from '@onaio/list-view';
import React, { Component, Fragment } from 'react';
import { Dropdown, DropdownItem, DropdownMenu, DropdownToggle, Row } from 'reactstrap';
import {
  ANC_REPORT,
  BIRTH_REPORT,
  CHILD_HEIGHT,
  CHILD_HEIGHT_MONITORING,
  CHILD_WEIGHT,
  CHILD_WEIGHT_MONITORING,
  CURRENT_PREGNANCY,
  GESTATION_PERIOD,
  HEIGHT,
  KG,
  MOTHER_WEIGHT_TRACKING,
  MOTHERS_WEIGHT,
  NUTRITION_REGISTRATION,
  NUTRITION_REPORT,
  PREGNANCY_REGISTRATION,
  WEIGHT,
} from '../../constants';
import { getNumberSuffix } from '../../helpers/utils';
import { SmsData } from '../../store/ducks/sms_events';
import WeightAndHeightChart from '../WeightAndHeightChart';
import './index.css';

interface Props {
  singlePatientEvents: SmsData[];
  isNutrition: boolean;
}

/**
 * An object that represents the weight of a child
 * or mother for a given month and year.
 * @member {number} weight  the weight
 * @member {number} month a number between 0 and 11 representing the month
 * @member {number} year the year the month is in.
 */
export interface WeightMonthYear {
  weight: number;
  month: number;
  year: number;
}

/**
 * An object representing Pregnancy data contained in an SmsEvent
 * object but has been adapted specifically for the pregnancy module
 * @member {string} EventDate - the date the SmsEvent was created. this is
 * when the sms was sent.
 * @member {string | number} message - the string/number representation of the message
 * @member {string} health_workder_name - name of health worker who sent the message
 * @member {string}  sms_type - the sms type.
 */
interface PregnancySmsData {
  EventDate: string;
  message: string | number;
  health_worker_name: string;
  sms_type: string;
}
interface State {
  pregnancyEventsArray: PregnancySmsData[][];
  dropdownOpenPregnancy: boolean;
  currentPregnancy: number;
  pregnancyDropdownLabel: string;
  indicesToRemove: number[];
}

export const convertToStringArray = (smsData: PregnancySmsData): string[] => {
  const arr: string[] = [];
  arr.push(smsData.sms_type);
  arr.push(smsData.EventDate);
  arr.push(smsData.health_worker_name);
  arr.push(smsData.message as string);
  return arr;
};

export const getEventsPregnancyArray = (
  singlePatientEvents: SmsData[],
  isNutrition: boolean
): PregnancySmsData[][] => {
  // remove event types that we are not interested in and retain
  // only pregnancy registration, ANC and birth reports
  singlePatientEvents = !isNutrition
    ? singlePatientEvents.filter((event: SmsData) => {
        return (
          event.sms_type.toLowerCase() === BIRTH_REPORT.toLowerCase() ||
          event.sms_type.toLowerCase() === ANC_REPORT.toLowerCase() ||
          event.sms_type.toLowerCase() === PREGNANCY_REGISTRATION.toLowerCase()
        );
      })
    : singlePatientEvents.filter((event: SmsData) => {
        return (
          event.sms_type.toLowerCase() === NUTRITION_REPORT.toLowerCase() ||
          event.sms_type.toLowerCase() === NUTRITION_REGISTRATION.toLowerCase()
        );
      });
  const data: PregnancySmsData[][] = [];

  let pregnancyIndex: number = 0;
  for (const dataItem in singlePatientEvents) {
    if (singlePatientEvents) {
      if (data[pregnancyIndex]) {
        if (
          singlePatientEvents[dataItem].sms_type === PREGNANCY_REGISTRATION ||
          (data[pregnancyIndex][parseInt(dataItem, 10) - 1] &&
            GESTATION_PERIOD <
              Date.parse(singlePatientEvents[dataItem].EventDate) -
                Date.parse(data[pregnancyIndex][parseInt(dataItem, 10) - 1].EventDate))
        ) {
          pregnancyIndex = pregnancyIndex + 1;
          if (!data[pregnancyIndex]) {
            data[pregnancyIndex] = [];
          }
          data[pregnancyIndex].push(singlePatientEvents[dataItem]);
        } else if (singlePatientEvents[dataItem].sms_type === BIRTH_REPORT) {
          data[pregnancyIndex].push(singlePatientEvents[dataItem]);
        } else {
          data[pregnancyIndex].push(singlePatientEvents[dataItem]);
        }
      } else {
        data[pregnancyIndex] = [];
        data[pregnancyIndex].push(singlePatientEvents[dataItem]);
      }
    }
  }

  return data;
};

class ReportTable extends Component<Props, State> {
  public static getDerivedStateFromProps(props: Props, state: State) {
    return {
      pregnancyEventsArray: getEventsPregnancyArray(props.singlePatientEvents, props.isNutrition),
    };
  }

  constructor(props: Readonly<Props>) {
    super(props);

    this.state = {
      currentPregnancy: 0,
      dropdownOpenPregnancy: false,
      indicesToRemove: [],
      pregnancyDropdownLabel: '',
      pregnancyEventsArray: [],
    };
  }

  public getPregnancyStringArray = (pregnancySmsData: PregnancySmsData[][]): string[][][] => {
    let pregnancySmsStrings: string[][][] = [];

    const gestation: number = GESTATION_PERIOD;
    for (const element in pregnancySmsData) {
      if (pregnancySmsData[element]) {
        pregnancySmsStrings[element] = pregnancySmsData[element].map(
          (sms: PregnancySmsData): string[] => {
            return convertToStringArray(sms);
          }
        );
      }
    }

    // filter out duplicate pregnancy registrations
    if (pregnancySmsStrings.length > 1) {
      for (const pregnancy in pregnancySmsStrings) {
        if (
          pregnancySmsStrings.length - 1 !== parseInt(pregnancy, 10) &&
          pregnancySmsStrings[parseInt(pregnancy, 10)].length === 1 &&
          gestation >
            Date.parse(pregnancySmsStrings[parseInt(pregnancy, 10)][0][1]) -
              Date.parse(pregnancySmsStrings[parseInt(pregnancy, 10) + 1][0][1]) &&
          pregnancySmsStrings[parseInt(pregnancy, 10)][0][0] === PREGNANCY_REGISTRATION &&
          pregnancySmsStrings[parseInt(pregnancy, 10) + 1][0][0] === PREGNANCY_REGISTRATION
        ) {
          this.state.indicesToRemove.push(parseInt(pregnancy, 10));
        }
      }
    }

    pregnancySmsStrings = pregnancySmsStrings.filter(
      (pregnancy, index) => !this.state.indicesToRemove.includes(index)
    );
    return pregnancySmsStrings;
  };

  public getWeightsArray = (
    pregnancySmsData: PregnancySmsData[][],
    field: string
  ): WeightMonthYear[][] => {
    let weights: WeightMonthYear[][] = [];
    for (const element in pregnancySmsData) {
      if (pregnancySmsData[element]) {
        weights[element] = pregnancySmsData[element].map(
          (sms: any): WeightMonthYear => {
            return {
              month: new Date(sms.EventDate).getMonth(),
              weight: sms[field],
              year: new Date(sms.EventDate).getFullYear(),
            };
          }
        );
      }
    }

    weights = weights.filter((pregnancy, index) => !this.state.indicesToRemove.includes(index));
    return weights;
  };

  public render() {
    const listViewProps = {
      data: this.getPregnancyStringArray(this.state.pregnancyEventsArray)[
        this.state.currentPregnancy
      ]
        ? this.getPregnancyStringArray(this.state.pregnancyEventsArray)[this.state.currentPregnancy]
        : [],
      headerItems: ['Report', 'Date', 'Reporter', 'Message'],
      tableClass: 'table-container',
      tbodyClass: 'body',
      tdClass: 'default-width',
    };

    return (
      <Fragment>
        <Row id="filter-panel">
          <p>Showing reports for:&emsp;</p>
          <div className="filters">
            <Dropdown
              isOpen={this.state.dropdownOpenPregnancy}
              toggle={this.togglePregnancyDropDown}
            >
              <DropdownToggle variant="success" id="dropdown-basic" caret={true}>
                <span>
                  {this.state.pregnancyDropdownLabel.length
                    ? this.state.pregnancyDropdownLabel
                    : 'select pregnancy'}
                </span>
              </DropdownToggle>
              <DropdownMenu>
                {this.getPregnancyStringArray(this.state.pregnancyEventsArray).map(
                  (pregnancy, i) => {
                    return (
                      <DropdownItem onClick={this.handlePregnancyDropDownClick} key={i}>
                        {(() => {
                          if (i === 0) {
                            return CURRENT_PREGNANCY;
                          } else {
                            const pregnancyIndex =
                              this.getPregnancyStringArray(this.state.pregnancyEventsArray).length -
                              i;
                            return pregnancyIndex + getNumberSuffix(pregnancyIndex) + ' pregnancy';
                          }
                        })()}
                      </DropdownItem>
                    );
                  }
                )}
              </DropdownMenu>
            </Dropdown>
          </div>
        </Row>
        <Row id="tableRow">
          <ListView {...listViewProps} />
        </Row>
        <Row id={'chart'}>
          <WeightAndHeightChart
            weights={
              this.getWeightsArray(this.state.pregnancyEventsArray, 'weight')[
                this.state.currentPregnancy
              ]
            }
            chartWrapperId={this.props.isNutrition ? 'nutrition-chart' : 'pregnancy-chart'}
            title={this.props.isNutrition ? CHILD_WEIGHT_MONITORING : MOTHER_WEIGHT_TRACKING}
            legendString={this.props.isNutrition ? CHILD_WEIGHT : MOTHERS_WEIGHT}
            units={KG}
            xAxisLabel={WEIGHT}
          />
          {this.props.isNutrition ? (
            <WeightAndHeightChart
              weights={
                this.getWeightsArray(this.state.pregnancyEventsArray, 'height')[
                  this.state.currentPregnancy
                ]
              }
              chartWrapperId={'nutrition-chart-1'}
              title={CHILD_HEIGHT_MONITORING}
              legendString={CHILD_HEIGHT}
              units={'cm'}
              xAxisLabel={HEIGHT}
            />
          ) : null}
        </Row>
      </Fragment>
    );
  }

  private togglePregnancyDropDown = () => {
    this.setState({
      dropdownOpenPregnancy: !this.state.dropdownOpenPregnancy,
    });
  };

  private handlePregnancyDropDownClick = (e: React.MouseEvent) => {
    // here we will take the new index and change the state using that index
    if ((e.target as HTMLInputElement).innerText === CURRENT_PREGNANCY) {
      this.setState({
        currentPregnancy: 0,
        pregnancyDropdownLabel: CURRENT_PREGNANCY,
      });
    } else {
      this.setState({
        currentPregnancy: parseInt((e.target as HTMLInputElement).innerText, 10),
        pregnancyDropdownLabel: (e.target as HTMLInputElement).innerText,
      });
    }
  };
}

export default ReportTable;
