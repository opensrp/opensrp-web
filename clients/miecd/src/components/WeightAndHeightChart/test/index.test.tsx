import { mount, render, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Highcharts from 'highcharts';
import { createBrowserHistory } from 'history';
import React from 'react';
import { Provider } from 'react-redux';
import { Router } from 'react-router';
import WeightAndHeightChart from '..';
import store from '../../../store/index';
import { WeightMonthYear } from '../../ReportTable';
import chartArgument from './fixtures';

jest.genMockFromModule('highcharts');
jest.mock('highcharts');

const history = createBrowserHistory();

const weights = [
    {
        month: 20,
        weight: 10,
        year: 2019,
    },
    {
        month: 8,
        weight: 10,
        year: 2019,
    },
    {
        month: 28,
        weight: 9,
        year: 2019,
    },
] as WeightMonthYear[];

const defaultProps = {
    chartWrapperId: 'wrapper-id',
    legendString: 'legend string',
    title: 'the title',
    units: 'kgs',
    xAxisLabel: 'x axis label',
};

jest.useFakeTimers();

describe('WeightAndHeightChart', () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });
    it('must render without crashing', () => {
        shallow(<WeightAndHeightChart weights={weights} {...defaultProps} />);
    });

    it('must render correctly', () => {
        const wrapper = mount(
            <Provider store={store}>
                <Router history={history}>
                    <WeightAndHeightChart weights={weights} {...defaultProps} />
                </Router>
            </Provider>,
        );
        expect(toJson(wrapper.find('#wrapper-id'))).toMatchSnapshot();
        wrapper.unmount();
    });

    it('calls hicharts.chart with the correct arguments', async () => {
        const wrapper = mount(
            <Provider store={store}>
                <Router history={history}>
                    <WeightAndHeightChart weights={weights} {...defaultProps} />
                </Router>
            </Provider>,
        );
        jest.runAllTimers();
        const argument1 = (Highcharts.chart as jest.MockedFunction<any>).mock.calls[0][0];
        const argument2 = (Highcharts.chart as jest.MockedFunction<any>).mock.calls[0][1];
        expect(argument1).toEqual('wrapper-id');
        (chartArgument.tooltip as any).formatter = argument2.tooltip.formatter;
        expect(typeof argument2.tooltip.formatter).toBe('function');
        expect(argument2).toEqual(chartArgument);
        wrapper.unmount();
    });
});
