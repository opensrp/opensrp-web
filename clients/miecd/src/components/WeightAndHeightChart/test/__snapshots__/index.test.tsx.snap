// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeightAndHeightChart must render correctly 1`] = `
<div
  id="wrapper-id"
/>
`;
